# Scripts for the Publication "Uncovering high levels of genomic diversity in the Southern Cone of South America"

This repository contains scripts related to the analysis conducted for the publication "Uncovering high levels of genomic diversity in the Southern Cone of South America" in XXX. The publication can be accessed XXX


## Data Availability

Due to the sensitive nature of the study, the raw data is not publicly available. However, it can be accessed by requesting access through this repository. The data have been deposited in XXX with accession number XXX.

## Overview

This repository hosts the scripts used for data analysis and visualization presented in the aforementioned publication. The scripts are organized into different folders based on their functionality and relevance to the study.


## Branches

You can navigate different content by switching branches. Each branch may represent a different aspect of the analysis or provide access to various versions of the scripts used in the study.
The data alignment was perfomed using this [pipeline](https://github.com/YourePrettyGood/fastq_to_gvcf_nextflow). The joint genotyping and calling with this [one](https://github.com/YourePrettyGood/gvcf_to_annotated_vcf_nextflow) and the data filter using this other [one](https://github.com/YourePrettyGood/PIBv1_manuscript).

- Figure 1 and related Figure (KING, PCA, mismatch count between array and WG, ROH) (https://github.com/epifaniarango/popgen_with_epi/tree/IBD)
- ALDER for dating admixture times (https://github.com/epifaniarango/popgen_with_epi/tree/ALDER)
- Masking ancestries in admixed populations (https://github.com/epifaniarango/popgen_with_epi/tree/Local-ancestry-and-masking)
- Dstats plots (https://github.com/epifaniarango/popgen_with_epi/tree/Dstats-plots)


## Contact Information

For any questions or further information, please contact epifaniarango@gmail.com .
 
